src/server/oauth/token.ts

```diff
@@ -0,0 +1,341 @@
+            tokens: authCode.tokens,
+            expiresAt: Math.floor((Date.now() + config.oauth.refreshTokenTimeoutMs) / 1000),
+            tableauClientId: authCode.tableauClientId,
+          });
+
+          setLongTimeout(
+            () => refreshTokens.delete(refreshTokenId),
+            config.oauth.refreshTokenTimeoutMs,
+          );
+
+          authorizationCodes.delete(code);
+
+          res.json({
+            access_token: accessToken,
+            token_type: 'Bearer',
+            expires_in: config.oauth.accessTokenTimeoutMs / 1000,
+            refresh_token: refreshTokenId,
+          });
+          return;
+        }
+        case 'client_credentials': {
+          // Generate access token for client credentials grant type.
+          // Refresh token is not supported for client credentials grant type.
+          // https://www.rfc-editor.org/rfc/rfc6749#section-4.4.3
+          const accessToken = await createClientCredentialsAccessToken(
+            {
+              clientId: clientCredentialsResult.value.clientId,
+              server: config.server,
+            },
+            publicKey,
+          );
+
+          res.json({
+            access_token: accessToken,
+            token_type: 'Bearer',
+            expires_in: config.oauth.accessTokenTimeoutMs / 1000,
+          });
+          return;
+        }
+        case 'refresh_token': {
+          // Handle refresh token
+          const { refreshToken } = result.data;
+          const tokenData = refreshTokens.get(refreshToken);
+          if (!tokenData || tokenData.expiresAt < Math.floor(Date.now() / 1000)) {
+            // Refresh token is expired
+            refreshTokens.delete(refreshToken);
+            res.status(400).json({
+              error: 'invalid_grant',
+              error_description: 'Invalid or expired refresh token',
+            });
+            return;
+          }
+
+          let accessToken: string;
+          const { refreshToken: tableauRefreshToken } = tokenData.tokens;
+
+          const tokensResult = await exchangeRefreshToken(
+            tokenData.server,
+            tableauRefreshToken,
+            tokenData.tableauClientId,
+          );
+
+          if (tokensResult.isErr()) {
+            // If the refresh token exchange fails, reuse the existing Tableau access token
+            // which may or may not be expired.
+            accessToken = await createAccessToken(
+              {
+                user: tokenData.user,
+                clientId: tokenData.clientId,
+                server: tokenData.server,
+                tokens: tokenData.tokens,
+              },
+              publicKey,
+            );
+          } else {
+            const {
+              accessToken: newTableauAccessToken,
+              refreshToken: newTableauRefreshToken,
+              expiresInSeconds,
+            } = tokensResult.value;
+
+            accessToken = await createAccessToken(
```

Comment:
We need to rotate the refresh token here.

See https://modelcontextprotocol.io/specification/2025-06-18/basic/authorization#token-theft